apps/system/js/audio_channel_manager.js

```diff
@@ -0,0 +1,370 @@
+     * put it into the `_interruptedAudioChannels` map.
+     * If it would like to be resumed when other audio channel ends,
+     * put it into `_audioChannelsResumeWhenOtherEnds` array.
+     *
+     * @param {AudioChannelController} audioChannel The audio channel.
+     * @param {Boolean} isResumedWhenOtherEnds
+     * Resume when other audio channel ends if as true.
+     */
+    _interruptAudioChannel: function(audioChannel, isResumedWhenOtherEnds) {
+      return audioChannel.pause().then(() => {
+        return new Promise((resolve, reject) => {
+          var id = audioChannel.instanceID;
+          this._activeAudioChannels.delete(id);
+          var backgroundResumeIndex;
+          if (isResumedWhenOtherEnds) {
+            var length =
+              this._audioChannelsResumeWhenOtherEnds.push(audioChannel);
+            backgroundResumeIndex = length - 1;
+          }
+          this._interruptedAudioChannels.set(id, {
+            audioChannel: audioChannel,
+            backgroundResumeIndex: backgroundResumeIndex
+          });
+          resolve();
+          this.debug('Interrupted ' + audioChannel.channelName);
+          this.debug('Resume when others end: ' + isResumedWhenOtherEnds);
+        });
+      });
+    },
+
+    /**
+     * Set the audio channel as default state as muted,
+     * and remove it from `_activeAudioChannels` map.
+     *
+     * @param {AudioChannelController} audioChannel The audio channel.
+     */
+    _resetAudioChannel: function(audioChannel) {
+      return audioChannel.pause().then(() => {
+        return new Promise((resolve, reject) => {
+          this._activeAudioChannels.delete(audioChannel.instanceID);
+          resolve();
+        });
+      });
+    },
+
+    /**
+     * Fade in all faded out audio channels,
+     * if the notification is ended.
+     *
+     * @param {String} channelName The name of the audio channel is just ended.
+     * @return {Promise}
+     */
+    _fadeInFadedOutAudioChannel: function(channelName) {
+      var promise;
+      var promises = [];
+      if (channelName === 'notification') {
+        this._activeAudioChannels.forEach((audioChannel) => {
+          audioChannel.states.fadingOut &&
+            promises.push(audioChannel.fadeIn());
+        });
+        promise = Promise.all(promises);
+      }
+      return promise;
+    },
+
+    /**
+     * Resume the audio channels in foreground.
+     *
+     * @param {AppWindow} app The app window in foreground.
+     * @return {Promise}
+     */
+    _resumeForegroundAudioChannels: function(app) {
+      var promise;
+      this._interruptedAudioChannels.size &&
+        app.browser.element.allowedAudioChannels.forEach((audioChannel) => {
+          var id = app.getAudioChannel(audioChannel.name).instanceID;
+          var element = this._interruptedAudioChannels.get(id);
+          if (element) {
+            promise = this._manageAudioChannels(element.audioChannel)
```

Comment:
Had bugs here.
The value of `promise` might be just overwrited for many times in the loop.
`promise = this._manageAudioChannels(element.audioChannel)`.